File: prime_finder.js
// prime_finder.js
function sieveBasic(limit) {
    const isPrime = new Array(limit + 1).fill(true);
    isPrime[0] = isPrime[1] = false;
    
    for (let i = 2; i * i <= limit; i++) {
        if (isPrime[i]) {
            for (let j = i * i; j <= limit; j += i) {
                isPrime[j] = false;
            }
        }
    }
    
    const primes = [];
    for (let i = 2; i <= limit; i++) {
        if (isPrime[i]) {
            primes.push(i);
        }
    }
    return primes;
}

function segmentedSieve(limit) {
    const sqrtLimit = Math.floor(Math.sqrt(limit)) + 1;
    const basePrimes = sieveBasic(sqrtLimit);
    
    let count = basePrimes.length;
    const segmentSize = 1000000;
    
    for (let low = sqrtLimit + 1; low <= limit; low += segmentSize) {
        const high = Math.min(low + segmentSize - 1, limit);
        const segment = new Array(high - low + 1).fill(true);
        
        for (const p of basePrimes) {
            let start = Math.floor((low + p - 1) / p) * p;
            if (start < low) start += p;
            if (start === p) start += p;
            
            for (let j = start; j <= high; j += p) {
                segment[j - low] = false;
            }
        }
        
        count += segment.reduce((sum, val) => sum + (val ? 1 : 0), 0);
    }
    
    return count;
}

const start = process.hrtime.bigint();
const result = segmentedSieve(10000000000);
const end = process.hrtime.bigint();

console.log(`Primes up to 10^10: ${result}`);
console.log(`Time: ${Number(end - start) / 1000000000} seconds`);

File: prime_finder.py
# prime_finder.py
import sys
import time
import math

def sieve_basic(limit):
    is_prime = [True] * (limit + 1)
    is_prime[0] = is_prime[1] = False
    
    for i in range(2, int(math.sqrt(limit)) + 1):
        if is_prime[i]:
            for j in range(i * i, limit + 1, i):
                is_prime[j] = False
    
    return [i for i in range(2, limit + 1) if is_prime[i]]

def segmented_sieve(limit):
    sqrt_limit = int(math.sqrt(limit)) + 1
    base_primes = sieve_basic(sqrt_limit)
    
    count = len(base_primes)
    segment_size = 1000000
    
    for low in range(sqrt_limit + 1, limit + 1, segment_size):
        high = min(low + segment_size - 1, limit)
        segment = [True] * (high - low + 1)
        
        for p in base_primes:
            start = ((low + p - 1) // p) * p
            if start < low:
                start += p
            if start == p:
                start += p
            
            for j in range(start, high + 1, p):
                segment[j - low] = False
        
        count += sum(segment)
    
    return count

if __name__ == "__main__":
    start = time.time()
    result = segmented_sieve(10**10)
    end = time.time()
    
    print(f"Primes up to 10^10: {result}")
    print(f"Time: {end - start:.2f} seconds")
    sys.exit(0) 